## Review: Bounce-Back Cycle 2

**Reason:** CI failed (unrelated, will rebase)

**Reviewer A:** skipped — CI red gates the merge regardless of reviewer verdict; reviewers will be re-run after the next push.
**Reviewer B:** skipped — same reason.
**CI:** red (4 FAILURE / 30 SUCCESS)

**Failed checks (all in workflow run [25952747817](https://github.com/stellar-experimental/henyey/actions/runs/25952747817)):**

- `build / 3 test / 2 test (testing-with-pr, amd64, local, rpc , )`
- `build / 3 test / 2 test (testing-with-pr, amd64, local, core,rpc,horizon , )`
- `build / 3 test / complete`
- `build / complete`

**Diff-attribution:** unrelated. The PR diff is two Rust files (`crates/history/src/catchup/download.rs`, `crates/history/src/catchup/persist.rs`) — a `TempDir::keep()` leak fix in test helpers. The failing checks are `Quickstart` `testing-with-pr` jobs that hang on `stellar-rpc` failing to reach ready state (repeated `WARN exited: stellar-rpc (exit status 1)` then "Run friendbot test has timed out after 4 minutes"). The same `Quickstart` workflow is currently failing on `main` (run on `2026-05-16T04:15:56Z`), so this is upstream infra flakiness, not caused by this PR.

**Failed-check excerpt:**

```
2026-05-16T04:46:00.0762634Z 2026/05/16 04:46:00 [test_friendbot] Waiting for Friendbot to be available
2026-05-16T04:46:01.5166432Z stellar-rpc: waiting for ready state, 4 minutes...
2026-05-16T04:46:08.3766780Z supervisor: 2026-05-16 04:46:08,376 WARN exited: stellar-rpc (exit status 1; not expected)
2026-05-16T04:46:27.9956207Z ##[error]The action 'Run friendbot test' has timed out after 4 minutes.
```

Routing back to `ready-for-doing` for `/do` Mode B. Recommended action: rebase on `origin/main` and re-push to trigger fresh CI; if `Quickstart` keeps failing on `main` itself, the next `/review-pr` will mark `blocked`.

--- a/crates/history/src/catchup/download.rs
+++ b/crates/history/src/catchup/download.rs
@@ -604,17 +604,30 @@ mod tests {
     use henyey_bucket::BucketManager;
     use henyey_db::{queries::StateQueries, Database};
 
-    fn make_test_catchup_manager() -> CatchupManager {
+    /// Returns the `TempDir` guard backing the `BucketManager` together
+    /// with the manager. The `TempDir` is returned **first** so callers
+    /// destructure as `let (_tmp_dir, manager) = ...`: Rust drops local
+    /// bindings in reverse declaration order, so binding `manager`
+    /// second guarantees it (and any file handles its `BucketManager`
+    /// holds) is dropped before the `TempDir` removes the directory.
+    /// The caller must keep `_tmp_dir` alive for the duration of the
+    /// test so the bucket directory is deleted on drop rather than
+    /// leaked under the system temp location.
+    fn make_test_catchup_manager() -> (tempfile::TempDir, CatchupManager) {
         let db = Database::open_in_memory().expect("in-memory db");
         let tmp_dir = tempfile::tempdir().expect("temp dir");
-        let bucket_manager = BucketManager::new(tmp_dir.keep()).expect("bucket manager");
+        let bucket_manager =
+            BucketManager::new(tmp_dir.path().to_path_buf()).expect("bucket manager");
         let archive = crate::HistoryArchive::new("https://example.com").expect("archive");
-        super::super::CatchupManager::new(vec![archive], bucket_manager, db)
+        (
+            tmp_dir,
+            super::super::CatchupManager::new(vec![archive], bucket_manager, db),
+        )
     }
 
     #[test]
     fn test_load_local_has_absent() {
-        let mgr = make_test_catchup_manager();
+        let (_tmp_dir, mgr) = make_test_catchup_manager();
         let result = mgr.load_local_has();
         assert!(result.is_ok(), "expected Ok, got: {:?}", result);
         assert!(result.unwrap().is_none(), "expected None for fresh DB");
@@ -622,7 +635,7 @@ mod tests {
 
     #[test]
     fn test_load_local_has_valid() {
-        let mgr = make_test_catchup_manager();
+        let (_tmp_dir, mgr) = make_test_catchup_manager();
         // Structurally valid v1 HAS with BUCKET_LIST_LEVELS zero-hash levels.
         let zero = "0".repeat(64);
         let zero_level = format!(
@@ -652,7 +665,7 @@ mod tests {
 
     #[test]
     fn test_load_local_has_corrupt_json() {
-        let mgr = make_test_catchup_manager();
+        let (_tmp_dir, mgr) = make_test_catchup_manager();
         mgr.db
             .with_connection(|conn| {
                 conn.set_state(

--- a/crates/history/src/catchup/persist.rs
+++ b/crates/history/src/catchup/persist.rs
@@ -222,12 +222,25 @@ mod tests {
         NetworkId::from_passphrase("Test SDF Network ; September 2015")
     }
 
-    fn make_test_catchup_manager() -> CatchupManager {
+    /// Returns the `TempDir` guard backing the `BucketManager` together
+    /// with the manager. The `TempDir` is returned **first** so callers
+    /// destructure as `let (_tmp_dir, manager) = ...`: Rust drops local
+    /// bindings in reverse declaration order, so binding `manager`
+    /// second guarantees it (and any file handles its `BucketManager`
+    /// holds) is dropped before the `TempDir` removes the directory.
+    /// The caller must keep `_tmp_dir` alive for the duration of the
+    /// test so the bucket directory is deleted on drop rather than
+    /// leaked under the system temp location.
+    fn make_test_catchup_manager() -> (tempfile::TempDir, CatchupManager) {
         let db = Database::open_in_memory().expect("in-memory db");
         let tmp_dir = tempfile::tempdir().expect("temp dir");
-        let bucket_manager = BucketManager::new(tmp_dir.keep()).expect("bucket manager");
+        let bucket_manager =
+            BucketManager::new(tmp_dir.path().to_path_buf()).expect("bucket manager");
         let archive = crate::HistoryArchive::new("https://example.com").expect("archive");
-        CatchupManager::new(vec![archive], bucket_manager, db)
+        (
+            tmp_dir,
+            CatchupManager::new(vec![archive], bucket_manager, db),
+        )
     }
 
     fn make_test_envelope() -> TransactionEnvelope {
@@ -511,7 +524,7 @@ mod tests {
 
     #[test]
     fn test_persist_ledger_history_happy_path() {
-        let manager = make_test_catchup_manager();
+        let (_tmp_dir, manager) = make_test_catchup_manager();
         let network_id = test_network_id();
 
         let envelope = make_test_envelope();
@@ -554,7 +567,7 @@ mod tests {
 
     #[test]
     fn test_persist_ledger_history_empty_ledger() {
-        let manager = make_test_catchup_manager();
+        let (_tmp_dir, manager) = make_test_catchup_manager();
         let network_id = test_network_id();
 
         let data = LedgerData::new(
@@ -575,7 +588,7 @@ mod tests {
 
     #[test]
     fn test_persist_ledger_history_with_absent_entries() {
-        let manager = make_test_catchup_manager();
+        let (_tmp_dir, manager) = make_test_catchup_manager();
         let network_id = test_network_id();
 
         let header = make_header(300);
@@ -612,7 +625,7 @@ mod tests {
 
     #[test]
     fn test_persist_ledger_history_batch_rollback() {
-        let manager = make_test_catchup_manager();
+        let (_tmp_dir, manager) = make_test_catchup_manager();
         let network_id = test_network_id();
 
         let envelope = make_test_envelope();
@@ -656,7 +669,7 @@ mod tests {
 
     #[test]
     fn test_persist_and_load_ledger_header_round_trip() {
-        let manager = make_test_catchup_manager();
+        let (_tmp_dir, manager) = make_test_catchup_manager();
         let network_id = test_network_id();
 
         let header = make_header(600);

